### PR TITLE
recutils: patch to fix recins not working outside of $TMPDIR

### PR DESCRIPTION
**Pull request by loreb**
Details at https://git.savannah.gnu.org/cgit/recutils.git/commit/utils/recutl.c?id=86f662a8202408134a235572ec60141d3082f975
Will no longer be necessary once 1.9 is released; fixes bug #30546
Thanks to ericonr for all the assistance.

note to self: bootstrap.conf is not shipped
add *temporary* dependency on automake
adding gnulib pieces is a real chore...

And then the GNU tools (rightfully) decided that the manpages were out
of date, because a file has been modified, and insisted on running
help2man - I simply told pre_configure to update the timestamps of the
manpages.

#### General
- [ ] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [ ] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

### CODE REVIEW

File: srcpkgs/recutils/template [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
-hostmakedepends="pkg-config"
+# TODO remove automake in 1.9; only needed for copy-file
+hostmakedepends="automake pkg-config"
 makedepends="acl-devel libgcrypt-devel libuuid-devel libcurl-devel"
+pre_configure() {
+	# XXX horrible kludge to avoid help2man - remove in 1.9
+	touch man/*.1
+	# XXX ONLY needed for copy-file - remove in 1.9
+	aclocal && automake
+}
```
Comment (ericonr) [inline srcpkgs/recutils/template:15]: This should go below the information block, above `librec1_package`librec1_package

**Review by ericonr — COMMENTED** [review]
Any idea how close they are to a 1.9? :D

File: srcpkgs/recutils/template (hunk 1 above)
Comment (loreb) [reply srcpkgs/recutils/template]: Done and added the version check.

File: srcpkgs/recutils/template [hunk 2]
```diff
@@ -1,18 +1,31 @@
...
 distfiles="${GNU_SITE}/recutils/recutils-${version}.tar.gz"
 checksum=df8eae69593fdba53e264cbf4b2307dfb82120c09b6fab23e2dad51a89a5b193
+patch_args=-Np1
 
+pre_configure() {
+	if test x"$version.$revision" != x1.8.2 ; then
+		echo hopefully you can remove the automake dependency
+		echo and the pre_configure kludge
+		exit 1
+	fi
```
Comment (ericonr) [inline srcpkgs/recutils/template:23]: I don't this is necessary, you should trust the comments...

File: srcpkgs/recutils/template (hunk 1 above)
Comment (loreb) [reply srcpkgs/recutils/template]: About the size of the patch, I was thinking that you could, more or less, clone the gnulib repo, then check for every file that is quoted in the patch that it's there semi-automatically, something like
```
$ fgrep -C3 "$(cat somefile.c | sed 's/^/+/g)" huge.patch
```

The "-C3" is to see that the patch begins and ends with the files; what is left is a few lines and most of them are trivial,
like adding the new files to the makefile.

File: srcpkgs/recutils/template (hunk 2 above)
Comment (loreb) [reply srcpkgs/recutils/template:23]: Ok, it was just en enforceable reminder.
Btw, I just realized something - I did the "touch man/*.1" kludge because `xbps-query -Rvs help2man` showed nothing,
but I'm on my debian computer, and even `xbps-query -Rvs vim` shows nothing.
That being said, the manpages *should* stay unchanged, as this is just a bugfix with zero functionality change.

**Comment by loreb** [comment]
Somehow I thought the "briefly tested this PR" was part of the PR review, silly me...
Anyway, the way I see it there are a handful of sensible ways to deal with it:
1. wait for upstream, as apparently they believe the fix made it into 1.8 but it didn't (https://lists.gnu.org/archive/html/bug-recutils/2021-04/msg00001.html)
2. Just apply the recommended workaround of setting $TMPDIR
3. prepare the patch yourself
4. verify this patch
5. ask upstream if they can release 1.8.1 with the fix

Number four is not as insane as it sounds, as the bulk of the patch is importing files from gnulib one by one;
I've prepared a small script to mostly automate that, here it is (expn below):
```
#! /bin/sh
set -e
set -u
# keep the patches in $PWD
ls copyfile.patch
ls exdev.patch
test -e recutils-1.8.tar.gz ||
wget https://ftp.gnu.org/gnu/recutils/recutils-1.8.tar.gz
tar zxpf recutils-1.8.tar.gz
cp -r recutils-1.8 patched

# apply patches
(
cd patched
patch -p1 < ../copyfile.patch
patch -p1 < ../exdev.patch
)

# get devel version
test -e recutils-1.8.90.tar.gz ||
wget https://alpha.gnu.org/gnu/recutils/recutils-1.8.90.tar.gz
tar zxf recutils-1.8.90.tar.gz

# remove files that didn't change in the devel version
(
cd recutils-1.8.90
find . -type f | while read i ; do
if cmp -s "$i" ../recutils-1.8/"$i" ; then rm -f "$i"; fi
done
)

# remove files that were not changed by the patches
(
cd patched
find . -type f | while read i ; do
if cmp -s "$i" ../recutils-1.8/"$i" ; then rm -f "$i"; fi
done
)

# show diff against devel version
(
cd patched
find . -type f | while read i ; do
diff -u "$i" ../recutils-1.8.90/"$i" || true
done
)
```

What it does: get the latest stable release, create a patched copy of it, get the devel release, remove files that didn't change
between 1.8+patches and 1.8 (in case you want to look at something by hand), diff what is left (ie the files that changed) against the devel version.
This shows that, indeed, the patch is only big because it pulls the dependencies of copy_file;
what is left is changes to Makefile.am, which I didn't import because I only wanted ONE change (compare it to the original),
and bumping the copyright from 2020 to 2021.

<rant> If I were king of the world, gnulib would ship as gnulib.a with headers in /usr/include/gnulib </rant>

Imho that brings the amount of work required down to a more than reasonable level;
otherwise just reject it with "wontfix - patch too big, wait for upstream", nobody else complained after all,
and it wouldn't be troublesome for *me* to trust myself keeping a version that I patched :=)

**Comment by loreb** [comment]
Generating the patch is easy:
start with 1.8, add the patch that fixed the bug, notice that it doesn't compile because of a missing dependency, say safe_write();
import lib/safe-write.c, lib/safe-write.h, m4/safe-write.m4, patch Makefile.am to add the extra dependency,
notice that is still doesn't compile, repeat until it compiles (and runs).

99% of what is imported is either small functions for convenience (eg safe read/write just wraps read/write to avoid EINTR and EINVAL - apparently some systems return EINVAL is the requested size is too big) or to work around peculiarities in this or that operating system; each file comes with yet more dependencies of course, so I had to import eight modules to get the one I wanted, then patch Makefile.am accordingly.

Other than that there's an ifdef for mac, the reason being that the script I posted above uses the alpha version of recutils
while to generate the patch I used a git checkout of gnulib.

But yeah, I guess I'll just ask upstream and if they take too long I'll modify the commit message like you said.

**Comment by loreb** [comment]
After 10 days with no answer on the mailing list, since this only requires a single function to copy one file, how about shelling out to /bin/cp? eg
```
#include <sys/wait.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

static int qcopy_file_preserving(const char *src, const char *dst)
{
        int stat_loc;
        pid_t pid = fork();
        if (pid < 0) {
                perror("fork");
                return -1;
        }
        if (pid == 0) {
                execl("/bin/cp", "--", src, dst, (void *)0);
                exit(1);
        }
        if (pid != waitpid(pid, &stat_loc, 0)) {
                perror("waitpid");
                return -1;
        }
        if (!(WIFEXITED(stat_loc) && WEXITSTATUS(stat_loc) == 0)) {
                return -1;
        }
        return 0;
}
```

should work fine until upstream makes a new release, right?

**Comment by loreb** [comment]
It's been one month since I asked upstream and still no response... to recap:
1. the fix is small, and most importantly the patch is literally copied from upstream
2. unfortunately it pull in a stupid amount of tiny dependencies, nearly 2kloc
3. I've given up generating number 2 in automatically
4. one month with no response means upstream will wait until 1.9 is out

`qcopy_file_preserving` above replaces all those stupid dependencies in the megapatch, at the cost of calling an external process; I tested it a couple of times to check for silly mistakes, then ran "make check" and all tests passed.
It's not even part of the shared libraries, so there should be no concerns like "what if something is using librec.{a,so} in a chroot without /bin/cp" - unless I'm mistaken of course.

TLDR: if you think it's fine, I'll replace the megapatch with qcopy_file_preserving above + any suggestion/modification etc,
otherwise just tell me and I'll close the PR, no big deal and no more time wasted :)

**Comment by ericonr** [comment]
If you feel ok with doing it via execing `cp`, I think I'd rather we carry that, yes! Don't forget to do `cp --` ;)

**Comment by loreb** [comment]
Done; the patch declares the function as static, meaning that the patch should immediately fail to compile when 1.9 is released and qcopy_file_preserving is declared "extern int...", and slightly modifies the original patch by commenting out `#include "copy-file.h"` as everything is done in a single file.

If I screwed up anything let me know.

File: srcpkgs/recutils/patches/copyfile.patch [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
++{
++        int stat_loc;
++        pid_t pid = fork();
++        if (pid < 0) {
++                perror("fork");
++                return -1;
++        }
++        if (pid == 0) {
++                execl("/bin/cp", "--", src, dst, (void *)0);
++                exit(1);
```
Comment (ericonr) [inline srcpkgs/recutils/patches/copyfile.patch:22]:
```suggestion
+                _exit(1);
```

Calling `exit` in a forked child can lead to all sorts of weirdness ;)

File: srcpkgs/recutils/template [hunk 4]
```diff
@@ -1,18 +1,26 @@
...
 homepage="https://www.gnu.org/software/recutils/"
 distfiles="${GNU_SITE}/recutils/recutils-${version}.tar.gz"
 checksum=df8eae69593fdba53e264cbf4b2307dfb82120c09b6fab23e2dad51a89a5b193
+patch_args=-Np1
 
+pre_configure() {
+	# XXX horrible kludge to avoid help2man - remove in 1.9
+	touch man/*.1
+	# XXX ONLY needed for copy-file - remove in 1.9
+	aclocal && automake
```
Comment (ericonr) [inline srcpkgs/recutils/template:22]: Is this still necessary?

File: srcpkgs/recutils/patches/copyfile.patch [hunk 5]
```diff
@@ -0,0 +1,37 @@
+--- a/utils/recutl.c
```
Comment (ericonr) [inline srcpkgs/recutils/patches/copyfile.patch:1]: Add a comment on top explaining the purpose of this patch

File: srcpkgs/recutils/template (hunk 4 above)
Comment (loreb) [reply srcpkgs/recutils/template]: Good catch; help2man is still called, but aclocal/acmake are no longer needed

File: srcpkgs/recutils/patches/copyfile.patch (hunk 5 above)
Comment (loreb) [reply srcpkgs/recutils/patches/copyfile.patch:9]: Done, sorry :D

File: srcpkgs/recutils/patches/copyfile.patch (hunk 3 above)
Comment (loreb) [reply srcpkgs/recutils/patches/copyfile.patch]: Ouch, there are a couple of calls to `atexit`, good call!

**Comment by loreb** [comment]
Closing because https://github.com/void-linux/void-packages/pull/37184 made this unnecessary
